### PR TITLE
asys

### PR DESCRIPTION
**Pull request by Aurel300**
Implementation of [HXP-0010](https://github.com/HaxeFoundation/haxe-evolution/blob/master/proposals/0010-asys.md).

Note: this PR has been split into #8831 and #8832, in addition to the target-specific PRs:

 - [HashLink](https://github.com/HaxeFoundation/hashlink/pull/306)
 - [Neko](https://github.com/HaxeFoundation/neko/pull/201)
 - eval implementations will be added soon

---

### TODO ###

 - [ ] For eval the API exposed to Haxe should be made closer to the one exposed by HashLink – less logic in OCaml, more in shared code
 - [x] (eval) `sys.io.FileInput` and `sys.io.FileOutput` were loaded hackily, this was moved to `Std` imports - maybe that's not good enough?
 - [ ] (eval) `-dce no` is required for the tests not to hang
 - [ ] (eval) Thread and Mutex have to be ported and used

### CODE REVIEW

**Comment by Simn** [comment]
Getting this to work on Windows was a bit of an adventure, so I'll document here what I did:

## libuv

1. Install cygwin packages `libtool` and `automake`
2. Follow [libuv build instructions](https://github.com/libuv/libuv#build-instructions) inside a cygwin shell, but run the configure command as `./configure --host=x86_64-w64-mingw32` (in case of 32 bit you probably replace some random numbers in that host name).
3. On `make`, I had an error about unknown defines in `udp.c` and patched it like so:

```diff
diff --git a/src/win/udp.c b/src/win/udp.c
index 3daa55f..7c015cc 100644
--- a/src/win/udp.c
+++ b/src/win/udp.c
@@ -790,9 +790,9 @@ int uv__udp_set_source_membership6(uv_udp_t* handle,
   memcpy(&mreq.gsr_source, source_addr, sizeof(mreq.gsr_source));
 
   if (membership == UV_JOIN_GROUP)
-    optname = MCAST_JOIN_SOURCE_GROUP;
+    optname = IP_ADD_SOURCE_MEMBERSHIP;
   else if (membership == UV_LEAVE_GROUP)
-    optname = MCAST_LEAVE_SOURCE_GROUP;
+    optname = IP_DROP_SOURCE_MEMBERSHIP;
   else
     return UV_EINVAL;
```

4. `make install` installs to `usr/local` by default, which (at least in my setup) is not picked up by the mingw-gcc compiler. There's probably a more elegant solution, but I just ended up copying some files and directories to `usr\x86_64-w64-mingw32\sys-root\mingw`:

```
bin/libuv-1.dll
include/uv
include/uv.h
lib/pkgconfig/libuv.pc
lib/libuv.a
lib/libuv.dll.a
lib/libuv.la
```

## Haxe

1. Currently, this patch is required:

```diff
diff --git a/libs/uv/uv_stubs.c b/libs/uv/uv_stubs.c
index 5486ac4..5cb6b57 100644
--- a/libs/uv/uv_stubs.c
+++ b/libs/uv/uv_stubs.c
@@ -1012,9 +1012,9 @@ CAMLprim value w_dns_getaddrinfo(value loop, value node, value flag_addrconfig,
 	UV_ALLOC_REQ(req, uv_getaddrinfo_t, cb);
 	int hint_flags_u = 0;
 	if (Bool_val(flag_addrconfig))
-		hint_flags_u |= AI_ADDRCONFIG;
+		hint_flags_u |= 0;
 	if (Bool_val(flag_v4mapped))
-		hint_flags_u |= AI_V4MAPPED;
+		hint_flags_u |= 0;
 	int hint_family_u = AF_UNSPEC;
 	if (Int_val(hint_family) == 4)
 		hint_family_u = AF_INET;

```

2. With that, Haxe should compile.
3. Check out https://github.com/Aurel300/haxe-sys
4. Run `haxe --cwd tests build-eval.hxml`. It will probably hang, so just comment out tests until it stops hanging and fails properly.

File: std/asys/uv/UVErrorType.hx [hunk 1]
```diff
@@ -1,81 +1,81 @@
...
-	var ETXTBSY = -26; // "text file is busy"
-	var EXDEV = -18; // "cross-device link not permitted"
-	var UNKNOWN = -4094; // "unknown error"
-	var EOF = -4095; // "end of file"
-	var ENXIO = -6; // "no such device or address"
-	var EMLINK = -31; // "too many links"
-	var EHOSTDOWN = -64; // "host is down"
-	var EOTHER = 0;
+extern enum abstract UVErrorType(Int) {
+	var E2BIG; // "argument list too long"
```
Comment (Gama11) [inline std/asys/uv/UVErrorType.hx]: Why not turn thse line comments into doc comments? Otherwise, the info won't show up in IDE completion or dox.

Comment (Aurel300) [reply std/asys/uv/UVErrorType.hx]: Will do. The same strings are also used when stringifying in `haxe.Error`, but docstrings should be available too. I think I wanted a longer, more meaningful description for the errors, but that can be done later anyway.

File: std/haxe/Timer.hx [hunk 2]
```diff
@@ -22,6 +22,15 @@
 
 package haxe;
 
+#if (target.asys)
```
Comment (Gama11) [inline std/haxe/Timer.hx:25]: I think it's no longer necessary to put dot-defines in parens?

File: src/context/common.ml [hunk 3]
```diff
@@ -588,6 +592,11 @@ let init_platform com pf =
 		define com Define.Sys
 	end else
 		com.package_rules <- PMap.add "sys" Forbidden com.package_rules;
+	if com.config.pf_asys then begin
+		raw_define_value com.defines "target.asys" "true";
+		define com Define.Asys
```
Comment (Gama11) [inline src/context/common.ml:597]: Why have both `target.asys` and `asys`? I thought the defines not prefixed by `target.` are just for backwards compatibility.

Comment (Simn) [reply src/context/common.ml:597]: If we have `sys` we should also have `asys`, otherwise this is just one giant trap.

File: libs/uv/uv_stubs.c [hunk 4]
```diff
@@ -1230,6 +1232,64 @@ CAMLprim value w_pipe_write_handle(value handle, value data, value send_handle,
 	UV_SUCCESS_UNIT;
 }
 
+// ------------- MUTEXES -----------------------------------------------
```
Comment (Aurel300) [inline libs/uv/uv_stubs.c]: You didn't even make sure the headers have the same length.

File: libs/uv/uv_stubs.c [hunk 5]
```diff
@@ -1230,6 +1232,64 @@ CAMLprim value w_pipe_write_handle(value handle, value data, value send_handle,
...
+
+CAMLprim value w_mutex_init(value unit) {
+	CAMLparam1(unit);
+	UV_ALLOC_CHECK(handle, uv_mutex_t);
+	UV_ERROR_CHECK_C(uv_mutex_init(Mutex_val(handle)), free(Mutex_val(handle)));
+	if ((UV_HANDLE_DATA(Mutex_val(handle)) = alloc_data()) == NULL)
```
Comment (Aurel300) [inline libs/uv/uv_stubs.c]: No need to `alloc_data` for a mutex, there are no callbacks.

**Comment by Simn** [comment]
See #8831. Branch will be kept to experiment further, but we don't need an open long-term PR for this.

**Pull request closed without merging**